app/views/play/VoyagerUnitMap.vue

```diff
@@ -0,0 +1,211 @@
+    2. position, nextLevels, first from the classroom snapshot, if does not exist then from campaign snapshot
+    4. any other data from the campaign snapshot, but if doesnt exist in campaign any more then use the data in classroom snapshot
+    TODO: Update after clarification from Robin - i.e. if we want them to be in sync always, so that its easy to update classroom snapshot of levels;
+    or if we want specific properties to only exist in one of these as the single source of truth.
+    */
+    async buildLevelsDataForCourse () {
+      try {
+        const courseInstance = await api.courseInstances.get({courseInstanceID: this.courseInstanceId})
+        const courseId = this.courseId = courseInstance.courseID
+        const classroomId = courseInstance.classroomID
+        
+        // campaign snapshot of the levels
+        const existingCampaignLevels = this.campaignData.levels
+
+        // classroom snapshot of the levels for the course
+        const classroom = this.classroom = await api.classrooms.get({classroomID: classroomId})
+        const classroomCourseLevels =  _.find(classroom.courses, {_id: courseId}).levels
+
+        // get levels data for the levels in the classroom snapshot
+        const classroomCourseLevelsData = await api.classrooms.getCourseLevels({classroomID: classroomId, courseID: courseId})
+
+        for (let level of classroomCourseLevels) {
+          this.classroomLevelMap[level.original] = level
+        }
+        
+        let courseLevelsData = {}
+        for (let level of classroomCourseLevelsData) {
+          let original = level.original
+          if (existingCampaignLevels[original]) {
+            courseLevelsData[original] = existingCampaignLevels[original]
+          }
+          else {
+            // a level which has been removed from the campaign but is saved in the course
+            courseLevelsData[original] = level
+          }
+          // carry over position, nextLevels, first property stored in classroom course, if there are any
+          if (this.classroomLevelMap[original].position)
+            courseLevelsData[original].position = this.classroomLevelMap[original].position
+          if (this.classroomLevelMap[original].nextLevels)
+            courseLevelsData[original].nextLevels = this.classroomLevelMap[original].nextLevels
+          if (this.classroomLevelMap[original].first)
+            courseLevelsData[original].first = this.classroomLevelMap[original].first
+        }
+        // TODO: what if level was removed from campaign, and classroom level map doesnt contain positions, nextLevels, and first? Is it a valid scenario?
+        // If yes, then use the logic from CampaignView for setting the missing positions using the prev and next level positions
+        // Not sure how to set the .first and .nextLevels in that case.
+
+        return courseLevelsData
+      }
+      catch (err) {
+        return Promise.reject(err)
```

Comment:
If an await method fails, isn't `Promise.reject(err)` the default response?

Reply:
Yeah but if we don't catch it, wouldn't it result in an uncaught exception?

Reply:
I believe this will also result in an uncaught exception in promise. As we're not catching the "reject" here. We're unwrapping it and re-wrapping it. Not sure without testing. But async function returns a promise. And an error in that promise causes it to `reject`.

Reply:
Yeah I think you are right, and we are catching the error returned from here in the `created` method : https://github.com/codecombat/codecombat/pull/5281/files#diff-7629799da91ec72bc985d79e6a0e3fc8R70
Regarding async function always returning a rejected promise in case of error, is that true even if error occurs outside of await call, i.e. for example if this line causes an error: https://github.com/codecombat/codecombat/pull/5281/files#diff-7629799da91ec72bc985d79e6a0e3fc8R126

Reply:
I ran an experiment and you're right. Thanks for teaching me something new!
The warning about the unhandled error vanishes even if you re-wrap it as a rejection.
An alternative is adding a `.catch` to the function call. That also removes the warning.

Thank you!